### PR TITLE
testlapack,native,cgo: add benchmark for Dgeev

### PR DESCRIPTION
**Pull request by vladimir-ch**
In `cgo`:
```
$ LD_LIBRARY_PATH=$HOME/src/OpenBLAS CGO_LDFLAGS="-L$HOME/src/OpenBLAS -lopenblas" go test -v -run=NONE -bench=Dgeev -cpu=6,6,6,6 > ../cgo_dgeev.txt
```
In `native`:
```
$ go test -run=NONE -bench=Dgeev -cpu=6,6,6,6 > ../native_dgeev.txt
```
In `lapack`:
```
$ benchstat native_dgeev.txt cgo_dgeev.txt
name                      old time/op  new time/op   delta
DgeevAntisymRandomN3-6    9.29µs ± 0%   7.55µs ± 1%  -18.73%  (p=0.029 n=4+4)
DgeevAntisymRandomN4-6    12.7µs ± 0%   10.9µs ± 0%  -14.08%  (p=0.029 n=4+4)
DgeevAntisymRandomN5-6    18.8µs ± 0%   19.1µs ± 0%   +1.41%  (p=0.029 n=4+4)
DgeevAntisymRandomN10-6   51.7µs ± 0%   58.2µs ± 2%  +12.57%  (p=0.029 n=4+4)
DgeevAntisymRandomN50-6   2.22ms ± 0%   2.79ms ± 0%  +25.80%  (p=0.029 n=4+4)
DgeevAntisymRandomN100-6  14.4ms ± 0%   22.7ms ± 6%  +57.80%  (p=0.029 n=4+4)
DgeevAntisymRandomN200-6  88.1ms ± 0%  102.5ms ±17%     ~     (p=0.343 n=4+4)
DgeevCirculantN3-6        9.02µs ± 0%   7.08µs ± 0%  -21.53%  (p=0.029 n=4+4)
DgeevCirculantN4-6        13.0µs ± 0%   11.2µs ± 0%  -14.07%  (p=0.029 n=4+4)
DgeevCirculantN5-6        16.0µs ± 0%   13.6µs ± 0%  -14.93%  (p=0.029 n=4+4)
DgeevCirculantN10-6       47.6µs ± 0%   48.1µs ± 1%   +0.97%  (p=0.029 n=4+4)
DgeevCirculantN50-6       1.65ms ± 0%   1.85ms ± 0%  +12.34%  (p=0.029 n=4+4)
DgeevCirculantN100-6      11.9ms ± 0%   19.0ms ±20%  +59.41%  (p=0.029 n=4+4)
DgeevCirculantN200-6      77.2ms ± 0%   63.3ms ±13%  -18.00%  (p=0.029 n=4+4)
```

### CODE REVIEW

**Comment by vladimir-ch** [comment]
Interesting: if I run the benchmark against MKL, I get 2-4 error messages per one benchmark run:
```
Intel MKL ERROR: Parameter 6 was incorrect on entry to DGEEV.
```
In LAPACKE that corresponds to `lda` which is equal to `n` in the benchmarks. Given that there are so few messages compared to the number of calls to Dgeev in the benchmarks, it looks like something might be wrong in MKL's LAPACKE.

Can anyone reproduce this? Any ideas?

**Comment by btracey** [comment]
Here is what I have from OpenBLAS (running CGO_LDFLAGS="-L/Users/brendan/software/OpenBLAS -lopenblas" go test -run none  -bench .)

````
~/Documents/mygo$ benchcmp dgeevgo.txt dgeevc.txt 
BenchmarkDgeevAntisymRandomN3-8       12682         11253         -11.27%
BenchmarkDgeevAntisymRandomN4-8       21740         15056         -30.75%
BenchmarkDgeevAntisymRandomN5-8       29176         28143         -3.54%
BenchmarkDgeevAntisymRandomN10-8      74693         82964         +11.07%
BenchmarkDgeevAntisymRandomN50-8      2626260       6607223       +151.58%
BenchmarkDgeevAntisymRandomN100-8     16461578      47925812      +191.14%
BenchmarkDgeevAntisymRandomN200-8     100736862     225389143     +123.74%
BenchmarkDgeevCirculantN3-8           13803         13677         -0.91%
BenchmarkDgeevCirculantN4-8           23000         17953         -21.94%
BenchmarkDgeevCirculantN5-8           26514         22530         -15.03%
BenchmarkDgeevCirculantN10-8          69807         74286         +6.42%
BenchmarkDgeevCirculantN50-8          2058403       5653528       +174.66%
BenchmarkDgeevCirculantN100-8         13811837      41464902      +200.21%
BenchmarkDgeevCirculantN200-8         86946081      199349869     +129.28%
````

Are those numbers actually right? How is it possible we're twice as fast as Fortran?

**Comment by btracey** [comment]
Oh, also @vladimir-ch , you can use the -count flag. "go test -run none -bench . -count 6"

**Comment by kortschak** [comment]
Do we want to consider moving benchmarks to subtest structure? It breaks comparison to previous states (and significantly breaks 1.5 and 1.6), but it significantly simplifies code.

**Comment by vladimir-ch** [comment]
> you can use the -count flag. "go test -run none -bench . -count 6"

Thanks, I didn't know about that (or forgot).

> Do we want to consider moving benchmarks to subtest structure? It breaks comparison to previous states (and significantly breaks 1.5 and 1.6), but it significantly simplifies code.

I'd really like to use sub-benchmarks, I didn't use them here because of 1.5 and 1.6. AFAICT, we have no previous lapack benchmarks, so I'm not worried about that.

> Are those numbers actually right? How is it possible we're twice as fast as Fortran?

I think they are most likely right, it may come from us doing Dgemm in parallel. I added a benchmark with n=500 and it looks like for larger sizes we are losing (not to mention MKL, there we lose for any size). Still I think it is not bad at all.

I also figured out the error message from MKL: I was linking to `-lmkl_intel_ilp64` instead of
`-lmkl_intel_lp64`.

So the updated benchmark comparison against OpenBLAS and MKL is:

```
$ benchstat native_dgeev.txt openblas_dgeev.txt
name                      old time/op  new time/op  delta
DgeevAntisymRandomN3-6    9.66µs ± 1%  7.24µs ± 1%  -25.06%  (p=0.029 n=4+4)
DgeevAntisymRandomN4-6    13.2µs ± 0%  10.6µs ± 0%  -19.61%  (p=0.029 n=4+4)
DgeevAntisymRandomN5-6    19.6µs ± 0%  18.9µs ± 0%   -3.62%  (p=0.029 n=4+4)
DgeevAntisymRandomN10-6   52.1µs ± 0%  57.7µs ± 1%  +10.93%  (p=0.029 n=4+4)
DgeevAntisymRandomN50-6   2.19ms ± 0%  2.76ms ± 0%  +26.04%  (p=0.029 n=4+4)
DgeevAntisymRandomN100-6  14.3ms ± 0%  23.4ms ± 2%  +63.78%  (p=0.029 n=4+4)
DgeevAntisymRandomN200-6  87.5ms ± 0%  87.8ms ±33%     ~     (p=0.343 n=4+4)
DgeevAntisymRandomN500-6   1.23s ± 1%   0.77s ±13%  -37.03%  (p=0.029 n=4+4)
DgeevCirculantN3-6        9.28µs ± 0%  6.80µs ± 0%  -26.80%  (p=0.029 n=4+4)
DgeevCirculantN4-6        13.3µs ± 0%  10.9µs ± 0%  -18.17%  (p=0.029 n=4+4)
DgeevCirculantN5-6        16.6µs ± 0%  13.3µs ± 0%  -19.57%  (p=0.029 n=4+4)
DgeevCirculantN10-6       47.5µs ± 0%  47.9µs ± 1%   +0.99%  (p=0.029 n=4+4)
DgeevCirculantN50-6       1.63ms ± 0%  1.86ms ± 0%  +13.71%  (p=0.029 n=4+4)
DgeevCirculantN100-6      11.8ms ± 0%  19.7ms ±19%  +66.76%  (p=0.029 n=4+4)
DgeevCirculantN200-6      76.4ms ± 0%  58.5ms ± 9%  -23.49%  (p=0.029 n=4+4)
DgeevCirculantN500-6       1.06s ± 0%   0.60s ±20%  -43.60%  (p=0.029 n=4+4)
```
```
$ benchstat native_dgeev.txt mkl_dgeev.txt     
name                      old time/op  new time/op  delta
DgeevAntisymRandomN3-6    9.66µs ± 1%  6.93µs ± 1%  -28.30%  (p=0.029 n=4+4)
DgeevAntisymRandomN4-6    13.2µs ± 0%   8.8µs ± 2%  -33.10%  (p=0.029 n=4+4)
DgeevAntisymRandomN5-6    19.6µs ± 0%  13.6µs ± 1%  -30.62%  (p=0.029 n=4+4)
DgeevAntisymRandomN10-6   52.1µs ± 0%  30.8µs ± 0%  -40.77%  (p=0.029 n=4+4)
DgeevAntisymRandomN50-6   2.19ms ± 0%  0.81ms ± 0%  -63.01%  (p=0.029 n=4+4)
DgeevAntisymRandomN100-6  14.3ms ± 0%   4.2ms ± 0%  -70.55%  (p=0.029 n=4+4)
DgeevAntisymRandomN200-6  87.5ms ± 0%  21.6ms ± 0%  -75.33%  (p=0.029 n=4+4)
DgeevAntisymRandomN500-6   1.23s ± 1%   0.19s ± 0%  -84.13%  (p=0.029 n=4+4)
DgeevCirculantN3-6        9.28µs ± 0%  6.64µs ± 2%  -28.49%  (p=0.029 n=4+4)
DgeevCirculantN4-6        13.3µs ± 0%   9.0µs ± 1%  -32.13%  (p=0.029 n=4+4)
DgeevCirculantN5-6        16.6µs ± 0%  10.9µs ± 1%  -34.45%  (p=0.029 n=4+4)
DgeevCirculantN10-6       47.5µs ± 0%  27.4µs ± 0%  -42.20%  (p=0.029 n=4+4)
DgeevCirculantN50-6       1.63ms ± 0%  0.64ms ± 1%  -61.08%  (p=0.029 n=4+4)
DgeevCirculantN100-6      11.8ms ± 0%   3.4ms ± 0%  -71.28%  (p=0.029 n=4+4)
DgeevCirculantN200-6      76.4ms ± 0%  16.3ms ± 0%  -78.64%  (p=0.029 n=4+4)
DgeevCirculantN500-6       1.06s ± 0%   0.15s ± 1%  -85.38%  (p=0.029 n=4+4)
```

**Comment by btracey** [comment]
Well, here are the results for posterity 

````
brendan:~/Documents/mygo$ benchstat dgeevgo154.txt dgeevgotip.txt 
name                      old time/op  new time/op  delta
DgeevAntisymRandomN3-8    22.8µs ± 3%  17.5µs ± 8%  -23.43%  (p=0.008 n=5+5)
DgeevAntisymRandomN4-8    30.0µs ±18%  24.1µs ± 2%  -19.56%  (p=0.002 n=6+6)
DgeevAntisymRandomN5-8    41.1µs ±22%  33.1µs ± 1%  -19.47%  (p=0.002 n=6+6)
DgeevAntisymRandomN10-8    109µs ± 1%    89µs ± 3%  -18.28%  (p=0.004 n=5+6)
DgeevAntisymRandomN50-8   3.44ms ±10%  2.71ms ±18%  -21.23%  (p=0.004 n=6+6)
DgeevAntisymRandomN100-8  21.6ms ± 3%  18.5ms ± 5%  -14.64%  (p=0.002 n=6+6)
DgeevAntisymRandomN200-8   142ms ± 8%   107ms ± 2%  -24.76%  (p=0.002 n=6+6)
DgeevCirculantN3-8        21.4µs ±20%  17.6µs ±11%  -17.68%  (p=0.002 n=6+6)
DgeevCirculantN4-8        28.0µs ± 0%  25.7µs ± 3%   -8.38%  (p=0.004 n=5+6)
DgeevCirculantN5-8        34.4µs ± 3%  28.0µs ± 3%  -18.66%  (p=0.004 n=5+6)
DgeevCirculantN10-8        156µs ±11%    76µs ± 1%  -51.12%  (p=0.002 n=6+6)
DgeevCirculantN50-8       2.64ms ±33%  1.87ms ±12%  -29.00%  (p=0.002 n=6+6)
DgeevCirculantN100-8      17.8ms ± 4%  14.1ms ± 2%  -21.09%  (p=0.002 n=6+6)
DgeevCirculantN200-8       121ms ± 4%    89ms ± 2%  -26.49%  (p=0.002 n=6+6)
````

**Comment by btracey** [comment]
@vladimir-ch  Your benchmark is using MLK's Dgeev, right? Not just the BLAS implementation?

I thought OpenBLAS did matrix multiplication in parallel?

**Comment by vladimir-ch** [comment]
> Your benchmark is using MLK's Dgeev, right? Not just the BLAS implementation?

Yes, MKL's Dgeev.

> I thought OpenBLAS did matrix multiplication in parallel?

In fact, I don't know and can't check it right now.

**Comment by btracey** [comment]
Benchmarks you may be interested in seeing @randall77 @josharian

**Comment by vladimir-ch** [comment]
Changed the Dgeev benchmark to the sub-benchmark structure. PTAL

File: testlapack/dgeev.go [hunk 1]
```diff
@@ -733,3 +733,51 @@ func dgeevTestForAntisymRandom(n int, rnd *rand.Rand) dgeevTest {
...
+		wr := make([]float64, n)
+		wi := make([]float64, n)
+		work := make([]float64, 1)
+		impl.Dgeev(lapack.ComputeLeftEV, lapack.ComputeRightEV, n, nil, n, nil, nil, nil, n, nil, n, work, -1)
+		work = make([]float64, int(work[0]))
+		b.Run(bm.name, func(b *testing.B) {
+			for i := 0; i < b.N; i++ {
+				b.StopTimer()
+				copyGeneral(a, bm.a)
+				b.StartTimer()
```
Comment (btracey) [inline testlapack/dgeev.go]: Do your benchmarks actually work? I seem to remember you can't start and stop the timer within the loop.

Comment (vladimir-ch) [inline testlapack/dgeev.go]: I think they do. I didn't find anything about not using `StartTimer` and `StopTimer` within the loop, quite the contrary, they exist exactly for this use case.

**Comment by vladimir-ch** [comment]
I have split the benchmarks into their own files that are built based on `go1.7` tag. PTAL

**Pull request merged**